### PR TITLE
handle backgroud properly in libhttpserver

### PR DESCRIPTION
**Pull request by songgao**
see #kbfs for details

### CODE REVIEW

File: libhttpserver/server.go [hunk 1]
```diff
@@ -187,26 +188,46 @@ const portStart = 7000
...
-	s.server = libkb.NewHTTPSrv(
-		g, libkb.NewPortRangeListenerSource(portStart, portEnd))
-	// Have to start this first to populate the ServeMux object.
-	if err = s.server.Start(); err != nil {
+	if err = s.startServer(); err != nil {
 		return nil, err
 	}
-	s.server.Handle(requestPathRoot,
-		http.StripPrefix(requestPathRoot, http.HandlerFunc(s.serve)))
+	go s.monitorAppState()
```
Comment (strib) [inline libhttpserver/server.go]: This needs to be shutdown somewhere.

File: libhttpserver/server.go [hunk 2]
```diff
@@ -187,26 +188,46 @@ const portStart = 7000
...
+	}
+	s.server.Handle(requestPathRoot,
+		http.StripPrefix(requestPathRoot, http.HandlerFunc(s.serve)))
+	return nil
+}
+
+func (s *Server) monitorAppState() {
+	state := keybase1.AppState_FOREGROUND
+	for {
+		state = <-s.g.AppState.NextUpdate(&state)
```
Comment (strib) [inline libhttpserver/server.go]: On mobile will you be sharing the same `AppState` as the service?  Does that mean only one of you will get this update?  It looks like `AppState.NextUpdate()` takes a lock, so there can only be one caller at once.  Maybe this doesn't share a `g` with the service?

Comment (strib) [reply libhttpserver/server.go]: Oh I see, it actually doesn't block, and saves all the channels.  Ok nevermind.

Comment (songgao) [reply libhttpserver/server.go]: That’s actually a great point. I was rushing this in without thinking about that. Thanks for making sure!

File: libhttpserver/server.go [hunk 3]
```diff
@@ -187,26 +189,52 @@ const portStart = 7000
...
-	// Have to start this first to populate the ServeMux object.
-	if err = s.server.Start(); err != nil {
+	if err = s.startServer(); err != nil {
 		return nil, err
 	}
-	s.server.Handle(requestPathRoot,
-		http.StripPrefix(requestPathRoot, http.HandlerFunc(s.serve)))
+	ctx, cancel := context.WithCancel(context.Background())
+	go s.monitorAppState(ctx)
+	s.cancel = cancel
```
Comment (jzila) [inline libhttpserver/server.go]: these 3 lines could happen in `s.startServer`, which would allow you not to have to re-assign `s.cancel`.

**Pull request merged**